src/analysis/Grader.cpp

```diff
@@ -125,10 +126,16 @@ void Grader::fillToolchainResultsJSON() {
         // attacker, tracking pass count.
         size_t passCount = 0, testCount = 0;
         for (const auto& subpackages : testSet[attacker]) {
-          for (const std::unique_ptr<TestFile>& test : subpackages.second) {
+          for (const TestPair& testpair : subpackages.second) {
+            const std::unique_ptr<TestFile>& test = testpair.first;
+            // Poll while we wait for the result
+            // TODO this could probably be replaced with some sort of interrupt,
+            //  (and probably should be), but better this than no threads
+            while (!testpair.second.has_value())
```

Comment:
`std::future::wait` here, or `std::future::wait_for` if you want a time out and give up, though the tester should be guaranteed to return since it has its own timeout.

Technically, I think option is probably not thread safe, so `std::optional::has_value` might return true before the value is set. I believe `std::future` is purpose-built to handle this.